### ISSUE TITLE
"zfs share" command doesn't work for SMB shares

### ISSUE THREAD

**Issue by rbabchis** [openzfs/zfs#4999]
I'm testing on Ubuntu 16.04 with the included zol 0.6.5 packages.

**zfs share** and **zfs share -a** just don't work. **zfs set sharesmb=on pool/dataset works fine**. I discovered that the shares are listed in **/etc/dfs/sharetab** but don't work. If I reboot or anything and try to share them again with **zfs share**, it will not work. **net usershare** list displays nothing. And of course because of this all shares are lost on every reboot.

Despite a dataset having the sharesmb=on value set, it will not share with **zfs share -a** and if I use **zfs share pool/dataset** it says:  **cannot share 'pool/dataset': filesystem already shared**, when **net usershare list** shows nothing**. Samba is running and the shares appear in /etc/dfs/sharetab, not that that is helping me.

root@igor:~# zfs get sharesmb pool2/Media
NAME         PROPERTY  VALUE     SOURCE
pool2/Media  sharesmb  on        local
root@igor:~# zfs get sharesmb pool2/home
NAME        PROPERTY  VALUE     SOURCE
pool2/home  sharesmb  on        local

root@igor:~# zfs share -a
root@igor:~# net usershare list
root@igor:~# ls /var/lib/samba/usershares/
root@igor:~#

I've detailed a bunch of stuff at serverfault... if that's not good enough, I can bring some of it over here:

http://serverfault.com/questions/797938/ubuntu-xenial-16-04-zfs-share-command-doesnt-work-for-smb-shares

**Comment by FransUrbo** [comment]
Well known problem I'm afraid. The SMB implementation that's currently in ZoL is seriously broken! I've rewritten it completely, but there is no interest what so ever to getting it merged.

**Comment by dracwyrm** [comment]
I mount folders on ZFS as smb shares, so I could use this functionality. I currently use Samba config files, but that is not so reliable if the order of loading is changed somehow.

**Comment by rbabchis** [comment]
Thanks FransUrbo,

Wow what a shame! I hope you can get it merged. It doesn't make sense to leave it broken.

I'm (sort of) surprised Ubuntu took something broken like that and set it all up like it was working and pretended nothing was wrong... oh well, they have a bug report from me now. It makes you wonder about their other packages.

I remember giving up on NFS shares for a similar reason a long time ago too. It seems like these things should be easier to do than rewriting ZFS, creating kernel modules, etc... :(

How did it get broken? I was using this feature with the ZoL DEB package on Debian a few months ago, no issues. Is there a bug/discussion I can read so I don't have to ask questions? 

Thanks

**Comment by FransUrbo** [comment]
> Wow what a shame! I hope you can get it merged. It doesn't make sense to leave it broken.

Indeed, but it's not going to happen.

> I remember giving up on NFS shares for a similar reason a long time ago too.

Yeah, I rewrote that to :)

> How did it get broken?

I designed it completely wrong to begin with. I knew that USER
shares had serious limitations, but I couldn't get REGISTRY
shares to work.

Later turned out, that after I upgraded Samba, I could get it
to work, and therefor rewrote the SMB driver.

> I was using this feature with the ZoL DEB package on Debian a few months ago, no issues. Is there a bug/discussion I can read so I don't have to ask questions? 


Yeah, I'm the one packaging those packages and I included my iSCSI,
SMB and NFS patches in those packages.

We're now having the (same) discussion about including them into the
official Debian GNU/Linux packages or not.

http://lists.alioth.debian.org/cgi-bin/mailman/listinfo/pkg-zfsonlinux-devel


I can't remember where we've discussed this, but in essence, nobody
wants to merge any share* patches because "in the distant future"
ZED should deal with this.

And although I agree fully that that's probably a much better solution,
I think it's a disservice to our users not to have something that DO
work in the meantime.
--
As soon as you find a product that you really like,
they will stop making it.
- Wilson's Law

**Comment by kingneutron** [comment]
Ubuntu 14.04 ZFS with Samba is working just fine for me, in fact I'm doing some consulting business for backup servers where this is the main feature. ZFS+Samba has always been the "killer/must have app" for me.

@FransUrbo , do you have your own PPA repo? Might get some traction if you can get Slashdot and Reddit to do an article or two exposing what's going on...

**Comment by FransUrbo** [comment]
No, and I don't have time or interest to setup one. Use the official Debian GNU/Linux packages.

**Comment by rauar** [comment]
+1 to get your stuff merged until a viable alternative gets released.

**Comment by jumbi77** [comment]
Yes, +1 too!!
You definitely should upstream your work. i guess nobody knows when the "distant future" will (ever) be. In the meanwhile your patches are much welcome, IMO.

**Comment by behlendorf** [comment]
@rauar @jumbi77 these are the kinds of changes we could make once the `zfs_share` and `zfs_unshare` tests are enabled and working in the test suite.  Any help you could offer on that front would be welcome.  Direction on running the test suite can be found [here](https://github.com/zfsonlinux/zfs/wiki/Building-ZFS#running-zloopsh-and-zfs-testssh) and we've like to get the following suites enabled.

```
tests/zfs-tests/tests/functional/cli_root/zfs_share
tests/zfs-tests/tests/functional/cli_root/zfs_unshare
```

**Comment by jumbi77** [comment]
for reference: https://github.com/zfsonlinux/zfs/commit/2f71caf2d926249920d1b9162550c56715cc6461

**Comment by grizzlyfred** [comment]
I dont want to pollute this thread and am still searching... But... I am unsure if I should open a new issue and where.
My problem is, that I switched to arch b/c ubuntu 16.04 hat too many issues for me. Now the (felt) past decade "zfs share" never worked for me, and samba often was too slow compared to a nfs share e.g. to my backup box with 2xGig-E bonded. For now, all my attempts to make nfs-server.service wait until zpools are imported and then share them vie /etc/exports have utterly failed. (Maybe a target such as network-manager-wait-online or similar is needed?.
If I would now that instead I coul use zfs properties for sharing, that would be so cool. So PLEASE if FransUrbo fixed that, PLEASE include, it can't get worse, can it?

**Comment by ColinIanKing** [comment]
Ping. What is the current state of this issue? Any updates?  We've got users wondering if this has or will be fixed. https://bugs.launchpad.net/ubuntu/+source/zfs-linux/+bug/1615405

**Comment by kpande** [comment]
^ it's also potentially a configuration issue since it works on some systems and not others.

**Comment by behlendorf** [comment]
The groundwork is now largely in place to properly support this in a maintainable way.  But it's still going to take some additional development work to finish this functionality so it works reliably across distributions.

**Comment by jumbi77** [comment]
So maybe @FransUrbo can finish his work he already done? Or do you mean something else by 'groundwork'?

**Comment by ajthemacboy** [comment]
Waiting on this as well.

**Comment by solenskiner** [comment]
also waiting

**Comment by ColinIanKing** [comment]
Hi there, what's the current state of the development work for this bug?

**Comment by kpande** [comment]
@ColinIanKing as stated earlier, it's a system configuration error. you do not have anywhere for usershares to go.

**Comment by JonLaliberte** [comment]
Forgive my ignorance here, but I'm having the same issue as described by OP and just want to see if there is a workaround until the issue is fixed upstream.

I see the shares in /etc/dfs/sharetab and also in /var/lib/samba/usershares, but when I connect the client reports that there are no shares available. Was working well on 14.04, but isn't on 16.04 with the included zol packages.

**Comment by oramirite** [comment]
I've been looking into this and am a little bummed out to find that these properties are in such a half-implemented state. I'm working on developing a kind of immutable private cloud infrastructure for some clients, and I was hoping that this feature did some very basic config generation that I could bootstrap into my own Samba install. I see that this feature utilizes the "usershare" feature of Samba, but only when a local installation of Samba is detected. Unfortunately there is no actual configuration of the share available, so a dataset like pool/data/set/Name it ends up with a name like "pool_dataset_Name". So it would be great to be able to nest simple configuration options like share name in here.

**Comment by FibreFoX** [comment]
Can someone please tame that stale-bot?

**Comment by scineram** [comment]
Why?
